File: Src/Python/server_socket.py
from socket import *
from threading import *
import json

class server_socket(object):
	def __init__(self):
		self.tcp_sock = socket(AF_INET, SOCK_STREAM)
		self.udp_sock = socket(AF_INET, SOCK_DGRAM)
		self.addr = gethostbyname(gethostname())
		self.transac = []
		self.blocks = []
	
	'''
		udp socket의 주소와 포트 고정
	'''
	def udp_bind(self):
		self.udp_sock.bind(('', 12222))

	'''
		client로부터 요청이 들어오면 검증 후 쓰레드를 생성해 데이터 전송
		:param
		subject : 요청하는 데이터를 구별하기 위한 문자열
		chain_info : 체인과 관련된 데이터
		transaction_info : transaction 데이터
	'''
	def udp_server(self, subject, chain_info):
		while True:
			data, c_addr = self.udp_sock.recvfrom(100)
			data = data.decode()
			if "Profit_Cut?"+subject in data:
				self.udp_sock.sendto("Profit_OK".encode(), c_addr)
				t = Thread(target = self.tcp_send_data, args = (chain_info,))
				t.start()
			elif "Profit_Cut_transac"+subject in data:
				data = data.split('//')
				transaction = json.loads(data[1])
				t = Thread(target = self.recv_transac, args = (transaction,))
				t.start()
			elif "Profit_Cut_block"+subject in data:
				data = data.split('//')
				block = json.loads(data[1])
				t = Thread(target = recv_block, args = (block,))
				t.start()
	'''
		새로 전파된 trasaction을 리스트에 저장
		:param
		trasaction : 새로 전파된 transaction
	'''			
	def recv_transac(self, transaction):
		self.transac.append(transaction)

	'''
		새로 전파된 block을 리스트에 저장
		:param
		block : 새로 전파된 block
	'''
	def recv_block(self, block):
		self.blocks.append(block)


	'''
		tcp socket의 주소와 포트 고정
	'''
	def tcp_bind(self):
		self.tcp_sock.bind(('', 12223))
		self.tcp_sock.listen(0)
	
	'''
		client로부터 요청이 들어오면 데이터 전송
	'''
	def tcp_send_data(self, info):	
		client_sock, addr = self.tcp_sock.accept()
		client_sock.send(json.dumps(info))		


	'''
		소켓 종료
	'''
	def disconnect(self, sock):
		sock.close()

sock = server_socket()
sock.udp_bind()
sock.tcp_bind()
info = [["chain"], ["checking list"]]
sock.udp_server('vote', info)

File: Src/Java/server_socket.java
package votechain;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.util.ArrayList;
import java.util.HashMap;

import javax.crypto.NoSuchPaddingException;

public class server_socket implements Runnable{
	private vote_block vb;
	private vote_chain vc;
	private ServerSocket tcp_sock;
	private DatagramSocket udp_sock;
	private ArrayList<HashMap> transacs;
	private ArrayList<block> blocks;
	private ArrayList<Object> chains;
	private String subject;
	private HashMap<String, HashMap> attendances; 
	
	public ArrayList<HashMap> getTransacs() {
		return transacs;
	}

	public ArrayList<block> getBlocks() {
		return blocks;
	}

	public ArrayList<Object> getChains() {
		return chains;
	}

	public HashMap<String, HashMap> getAttendances() {
		return attendances;
	}

	public server_socket(vote_block vb, vote_chain vc, ServerSocket tcp_sock, DatagramSocket udp_sock, ArrayList<Object> chains, ArrayList<HashMap> transacs, String subject, HashMap attendances) throws IOException {
		super();
		this.vb = vb;
		this.vc = vc;
		this.tcp_sock = tcp_sock;
		this.udp_sock = udp_sock;
		this.transacs = transacs;
		this.subject = subject;
		this.blocks = new ArrayList<block>();
		this.chains = chains;
		this.attendances = attendances;
	}
	
	private void udp_server() throws IOException, ClassNotFoundException, NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeySpecException, InterruptedException {
		byte[] buf = new byte[512];
		DatagramPacket packet = new DatagramPacket(buf, buf.length);
		this.udp_sock.receive(packet);
		
		ObjectInputStream iStream = new ObjectInputStream(new ByteArrayInputStream(buf));
		HashMap data = (HashMap) iStream.readObject();
		iStream.close();
		
		InetAddress address = packet.getAddress();
		int port = packet.getPort();
		packet = new DatagramPacket(buf, buf.length, address, port);
		
		if(data.containsKey("Profit_Cut?"+subject)) {
			buf = "Profit_OK".getBytes();
			packet = new DatagramPacket(buf, buf.length, address, 12222);
			this.udp_sock.send(packet);
			tcp_send_data();
		}
		else if(data.containsKey("Profit_Cut_transac"+subject)) {
			HashMap transac = (HashMap)data.get("Profit_Cut_transac"+subject);
			getRsa rsa = new getRsa();
			String sender = (String) data.get("sender");
			String encrypt_num = (String) data.get("encrypted");
			PublicKey key = (PublicKey) this.attendances.get(sender).get("Key");
			if(!rsa.decryption(encrypt_num, key).equals("false")) {
				this.transacs.add(transac);
			}
		}
		else if(data.containsKey("Profit_Cut_block"+subject)) {
			block b = (block) data.get("Profit_Cut_block"+subject);
			if(vb.valid_block(b, this.chains)) {
				vc.add_block(b);
			}
		}
		else if(data.containsKey("Profit_Cut_newbie"+subject)) {
			String account = (String)data.get("Profit_Cut_newbie"+subject);
			String encrypted = (String)data.get("encrypted");
			getRsa rsa = new getRsa();
			PublicKey key = (PublicKey)(rsa.decode_publickey((String)(data.get("Key"))));
			String user = rsa.decryption(encrypted, key);
			HashMap<String, Object> info = new HashMap<>();
			info.put("Key", key);
			info.put("token", 1);
			if(!user.equals("false") && !this.attendances.containsKey(user)) {
				this.attendances.put(account, info);
			}
		}
	}
	
	private void tcp_send_data() throws IOException, InterruptedException {
		while(true) {
			Socket socket = this.tcp_sock.accept();
			System.out.println("accepted!");
			ThreadHandler handler = new ThreadHandler(socket, this.chains);
			handler.start();
			handler.join();
		}
	}
	
	private byte[] message_serialize(ArrayList<Object> chain) throws IOException {
		ByteArrayOutputStream bStream = new ByteArrayOutputStream();
		ObjectOutput oo = new ObjectOutputStream(bStream); 
		oo.writeObject(chain);
		oo.close();
		byte[] serializedMessage = bStream.toByteArray();
		
		return serializedMessage;
	}
	@Override
	public void run() {
		while(true) {
			try {
				udp_server();
			} catch (ClassNotFoundException | NoSuchAlgorithmException | NoSuchPaddingException | IOException | InvalidKeySpecException | InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
	
	class ThreadHandler extends Thread{
		private Socket connectedClientSocket;
		private ArrayList<Object> chain;

		public ThreadHandler(Socket connectedClientSocket, ArrayList<Object> chain) {
			super();
			this.connectedClientSocket = connectedClientSocket;
			this.chain = chain;
		}
		public void run() {
			try {
				byte[] chain = message_serialize(this.chain);
				OutputStream os = this.connectedClientSocket.getOutputStream();
				ObjectOutputStream oos = new ObjectOutputStream(os);
				oos.writeObject(chain);
				oos.flush();
			} catch(IOException ignored) {
			} finally {
				try {
					this.connectedClientSocket.close();
				} catch(IOException ignored) {}
			}
		}
	}
	public static void main(String[] args) {
}
}
	

